tokio-net/src/driver/reactor/sharded_slab/page/mod.rs

```diff
@@ -0,0 +1,299 @@
+            // free list onto the local free list.
+            self.remote_head.swap(Self::NULL, Ordering::Acquire)
+        };
+
+        // if the head is still null, both the local and remote free lists are
+        // empty --- we can't fit any more items on this page.
+        if head == Self::NULL {
+            #[cfg(test)]
+            test_println!("-> NULL! {:?}", head);
+            return None;
+        }
+
+        // do we need to allocate storage for this page?
+        let page_exists = self.slab.with(|s| unsafe { (*s).is_none() });
+        if page_exists {
+            self.fill();
+        }
+
+        let gen = self.slab.with(|slab| {
+            let slab = unsafe { &*(slab) }
+                .as_ref()
+                .expect("page must have been allocated to alloc!");
+            let slot = &slab[head];
+            local.set_head(slot.next());
+            slot.alloc()
+        });
+
+        let index = head + self.prev_sz;
+        #[cfg(test)]
+        test_println!("alloc at offset: {}; gen={:?}", index, gen);
+        Some(gen.pack(index))
+    }
+
+    #[inline]
+    pub(in crate::driver) fn get(&self, addr: Addr, idx: usize) -> Option<&ScheduledIo> {
+        let page_offset = addr.offset() - self.prev_sz;
+        #[cfg(test)]
+        test_println!("-> offset {:?}", page_offset);
+
+        let value = self.slab.with(|slab| {
+            unsafe { &*slab }
+                .as_ref()?
+                .get(page_offset)?
+                .get(slot::Generation::from_packed(idx))
+        });
+        #[cfg(test)]
+        test_println!("-> offset {:?}; value={:?}", page_offset, value);
+        value
+    }
+
+    pub(crate) fn remove_local(&self, local: &Local, addr: Addr, idx: usize) {
+        let offset = addr.offset() - self.prev_sz;
+
+        #[cfg(test)]
+        test_println!("-> offset {:?}", offset);
+
+        self.slab.with(|slab| {
+            let slot =
+                if let Some(slot) = unsafe { &*slab }.as_ref().and_then(|slab| slab.get(offset)) {
+                    slot
+                } else {
+                    return;
+                };
+            if slot.reset(slot::Generation::from_packed(idx)) {
+                slot.set_next(local.head());
+                local.set_head(offset);
+            }
+        })
+    }
+
+    pub(crate) fn remove_remote(&self, addr: Addr, idx: usize) {
+        let offset = addr.offset() - self.prev_sz;
+
+        #[cfg(test)]
+        test_println!("-> offset {:?}", offset);
+
+        self.slab.with(|slab| {
+            let slot =
+                if let Some(slot) = unsafe { &*slab }.as_ref().and_then(|slab| slab.get(offset)) {
+                    slot
+                } else {
+                    return;
+                };
+            if !slot.reset(slot::Generation::from_packed(idx)) {
+                return;
+            }
+
+            let mut next = self.remote_head.load(Ordering::Relaxed);
```

Comment:
Can you explain why using `Relaxed` is correct here?

Reply:
Looks like I also used `Relaxed`... i will need to investigate why `loom` did not complain: https://github.com/tokio-rs/tokio/blob/master/tokio-executor/src/task/stack.rs#L32

Reply:
Ok, I remember why I used `Relaxed` there... i will need to document it, because I forgot why initially and it is fairly specific to that use case.

Reply:
I'm also not saying it is incorrect here, but the reasoning should, at the very least, be documented as it is not obviously correct.

Reply:
@carllerche the `compare_exchange` that sets the new head is `AcqRel` on a success, and `Acquire` on a failure, so if a push occurs concurrently, we'll synchronize with that and fail, and the load (the failure of the `compare_exchange`) will have the `Acquire` ordering, so the current value for the next attempt will have been loaded with `Acquire`. I thought loading the initial value with `Relaxed` would allow us to avoid an `Acquire` if there are no concurrent writes.

Is that correct?

Reply:
based on our offline discussion + additional loom tests I've added, my understanding is that this is correct?